src/jace/jax/translation_cache.py

```diff
@@ -0,0 +1,297 @@
+            strides = getattr(val, "strides", None)
+            # TODO(phimuell): is `CPU_Heap` always okay? There would also be `CPU_Pinned`.
+            storage = (
+                dace.StorageType.GPU_Global if util.is_on_device(val) else dace.StorageType.CPU_Heap
+            )
+
+            return cls(shape=shape, dtype=dtype, strides=strides, storage=storage)
+
+        if util.is_scalar(val):
+            shape = ()
+            dtype = util.translate_dtype(type(val))
+            strides = None
+            # Lets pretend that scalars are always on the CPU, which is a fair assumption.
+            storage = dace.StorageType.CPU_Heap
+
+            return cls(shape=shape, dtype=dtype, strides=strides, storage=storage)
+
+        raise TypeError(f"Can not make 'an abstract description from '{type(val).__name__}'.")
+
+
+@runtime_checkable
+class _ConcreteCallArgument(Protocol):
+    """Type for encoding a concrete arguments in the cache."""
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        pass
+
+    @abstractmethod
+    def __eq__(self, other: Any) -> bool:
+        pass
+
+
+"""This type is the abstract description of a function call.
+It is part of the key used in the cache.
+"""
+CallArgsDescription: TypeAlias = tuple[
+    _AbstarctCallArgument
+    | _ConcreteCallArgument
+    | tuple[str, _AbstarctCallArgument]
+    | tuple[str, _ConcreteCallArgument],
+    ...,
+]
+
+
+@dataclass(init=True, eq=True, frozen=True)
+class CachedCallDescription:
+    """Represents the structure of the entire call in the cache and used as key in the cache.
+
+    This class represents both the `JaceWrapped.lower()` and `JaceLowered.compile()` calls.
+
+    The actual key is composed of two parts, first the "origin of the call".
+    For this we just use the address of the stage object we are caching and hope that the
+    address is not reused for another stag anytime soon.
+
+    The second part is of the key are a description of the actual arguments, see `CallArgsDescription` type alias.
+    There are two ways for describing the arguments:
+    - `_AbstarctCallArgument`: Which encode only the structure of the arguments.
+        These are essentially the tracer used by Jax.
+    - `_ConcreteCallArgument`: Which represents actual values of the call.
+        These are either the static arguments or compile options.
+
+    While `JaceWrapped.lower()` uses both, `JaceLowered.compile()` will only use concrete arguments.
+    In addition an argument can be positional or a named argument,
+    in which case it consists of a `tuple[str, _AbstarctCallArgument  | _ConcreteCallArgument]`.
+
+    Notes:
+        The base assumption is that the stages are immutable.
+
+    Todo:
+        - pytrees.
+        - Turn the references into week references, Jax does this and I am sure there is a reason for it.
+    """
+
+    stage_id: int
+    fargs: CallArgsDescription
+
+
+class TranslationCache:
```

Comment:
Why cannot use the standard `__getitem__`, `__setitem__`, `__delitem__` and `__contains__` operators instead of custom function (`get`, `add`, `_evict`, `has`) doing exactly the same?